### ISSUE TITLE
equivalent to pip install -e . ?

### ISSUE THREAD

**Issue by bmabsout** [DavHau/mach-nix#339]
Hello, is there an equivalent to calling `pip install -e .` in a virtual-env for example? I'm currently using a flake and having to call `flake update` every time while developing a package is cumbersome. Is the right way to handle this to actually include pip in the shell and call `pip install -e .` in the shell hook?

**Comment by YodaEmbedding** [comment]
Note that this is called an "editable install".

**Comment by DavHau** [comment]
> Hello, is there an equivalent to calling `pip install -e .` in a virtual-env for example? I'm currently using a flake and having to call `flake update` every time while developing a package is cumbersome. Is the right way to handle this to actually include pip in the shell and call `pip install -e .` in the shell hook?

Why would you need to update the flake each time? What input is it that needs to be updated constantly. It sounds like that you split something into different flakes that should actually be in one flake.

**Comment by bmabsout** [comment]
There are 2 things I would like to have and editable install for , there's a seperate repository that I am including as a submodule and there's the package I work on directly. Right now what I do with respect to mach-nix is I run everything from the main folder so the imports work (or I create a virtualenv and pip install -e in there). But with an editable install, the current package would be in scope at all times. If you look at the poetry2nix documentation, there's a "mkPoetryEditablePackage"

**Comment by Yajo** [comment]
So there's no way to use mach-nix to install editable package(s)?

FTR, Odoo cannot be installed properly with pip using other method: https://github.com/odoo/odoo/issues/16700

**Comment by TyberiusPrime** [comment]
> So there's no way to use mach-nix to install editable package(s)?

Exactly, it's basically out of scope (editable installs are kind-of-non-reproducible by design).

[anysnake2](https://github.com/TyberiusPrime/anysnake2) can combine mach-nix and editable installs - provided you're happy with your programs running in a singularity container.

**Comment by Yajo** [comment]
My objective is to set up a development environment, where I need to edit multiple dependencies simultaneously (basically: odoo + extra addons).

According to my tests, while developing, purity is an enemy. Development is dirty by definition. To let mach-nix to serve as a replacement for Pipenv, Poetry, PDM... it needs to support development.

Of course, for packaging the final product, I'd use another kind of install (more or less the same `requirements` argument, but removing some `-e` flags from there). AFAIK this is a normal workflow, as flakes let us split the `devShell` definition from `packages`.

FTR Python supports deploying `.pth` files which point to other on-disk paths, son an editable install could just produce that file in the nix store as a python module, which the python interpreter would resolve appropriately. That would be great for development. If I'm not wrong, [that's what poetry2nix does](https://github.com/nix-community/poetry2nix/blob/302a559a298ba39a0bec53f46479da9c5b2558cd/editable.nix#L35-L51).

**Comment by DavHau** [comment]
If someone wants to make a PR to add a feature equivalent to what poetry2nix does, I'm open for it

But why don't you just add your local 'editable' project source to PYTHONPATH in your dev shell. Doesn't that solve the problem?

**Comment by Yajo** [comment]
That will not install the dependencies. Also it would forbid things like
PEP660.

**Comment by DavHau** [comment]
> That will not install the dependencies. Also it would forbid things like PEP660.

To get the dependencies you can just build a shell with mach-nix with all the dependencies included and then add your project via PYTHONPATH ontop.
But yeah, I see that not all aspects of PEP660 are covered by that, like entry points for example.

**Comment by Yajo** [comment]
> To get the dependencies you can just build a shell with mach-nix with all the dependencies included and then add your project via PYTHONPATH ontop.

But that would mean duplicating the dependencies, right? Having them defined on the project itself and on mach-nix.

If the project uses a `requirements.txt` that's no problem, but what if it uses [poetry](https://python-poetry.org/), [pdm](https://pdm.fming.dev/), [Pipenv](https://pipenv.pypa.io/), or any [PEP517](https://www.python.org/dev/peps/pep-0517/) backend?

**Comment by TyberiusPrime** [comment]
> If the project uses a `requirements.txt` that's no problem, but what if it uses [poetry](https://python-poetry.org/), [pdm](https://pdm.fming.dev/), [Pipenv](https://pipenv.pypa.io/), or any [PEP517](https://www.python.org/dev/peps/pep-0517/) backend?

Welcome to the wonderful world of [xcd](https://xkcd.com/927/).

Just like to point out that PEP517 defines build-time requirements, not runtime requirements.

In anysnake2, I  read requirements.txt and setup.cfg ('manually'), and push that into mach nix's requirementsExtra.

It's a bit bare bones, but enough for my packages.

The install is then handled via pip -e by installing into a temporary venv (with all requirements being supplied by mach-nix), and hacking the entry-point scripts to point to the correct mach-nix environment.

It's not pretty, but it does work. They even work in a 'nix develop' shell afterwards without a container (it does use a container to build though).

**Comment by Yajo** [comment]
Couldn't we just pip install -e the dependency and force pip to output
things into the derivation's outPath? This way we would let pip handle any
standard it wants.

**Comment by bmabsout** [comment]
Correct me if I'm wrong but I thought if I had a flake I could not directly refer to a relative subfolder with a relative path and append it to PYTHONPATH since all of it has to be copied to the store first. The other option is to hard code the path to my system but that's a far cry from being reproducible. If I can just refer to the local path and add it to PYTHONPATH that would solve the issue but I'm not sure how I can achieve that either. My goal is to allow anyone to import the package from other files and modify it quickly as well. Commits would take care of reproducibility of the source and mach-nix takes care of the deps
